web/packages/teleterm/src/ui/services/clusters/clustersService.ts

```diff
@@ -71,11 +71,29 @@ export class ClustersService extends ImmutableStore<types.ClustersServiceState>
     return cluster;
   }
 
+  /**
+   * Logs out of the cluster and removes the profile.
+   * Does not remove the cluster from the state, but sets the cluster and its leafs as disconnected.
+   * It needs to be done, because some code can operate on the cluster the intermediate period between logout
+   * and actually removing it from the state.
+   * A code that operates on that intermediate state is in `useClusterLogout.tsx`.
+   * After invoking `logout()`, it looks for the next workspace to switch to. If we hadn't marked the cluster as disconnected,
+   * the method might have returned us the same cluster we wanted to log out of.
+   */
   async logout(clusterUri: uri.RootClusterUri) {
     // TODO(gzdunek): logout and removeCluster should be combined into a single acton in tshd
     await this.client.logout(clusterUri);
-    await this.removeCluster(clusterUri);
-    await this.removeClusterKubeConfigs(clusterUri);
+    await this.client.removeCluster(clusterUri);
+
+    this.setState(draft => {
+      draft.clusters.forEach(cluster => {
+        const isRootOrLeafThatBelongsToRoot =
+          routing.ensureRootClusterUri(cluster.uri) === clusterUri;
+        if (isRootOrLeafThatBelongsToRoot) {
```

Comment:
```suggestion
        if (routing.belongsToProfile(clusterUri, cluster.uri) {
```

Reply:
The name could be better, both in terms how it sounds overall and also because I don't think we widely use "profile" within Connect. I didn't have a better idea at the time though. Maybe something with "workspace". But the URI itself refers to a profile so idk.

Reply:
To me "profile" is fine :) For example, "workspace" seems like a higher level concept, much higher than the URIs.